### PR TITLE
Update adapter creation guide

### LINKED ISSUE #5104: [May 2] Update Adapter Creation Guide for v1.8 interface changes
The guide https://docs.getdbt.com/guides/adapter-creation?step=1 needs to be reviewed and updated based on the changes to the adapters interface (ie dbt-common + dbt-adapters).
- https://github.com/dbt-labs/dbt-core/discussions/9171

### When is this due?
This is needed for v1.8 (Core) + "Keep on latest version" (Cloud) and can be worked on at any time, but is due May 2nd for 1.8 final GA

related: https://github.com/dbt-labs/dbt-adapters/issues/136

### Changes

- [x] update https://docs.getdbt.com/guides/adapter-creation?step=2#keeping-the-code-compatible-with-dbt-core
- [x] update https://docs.getdbt.com/guides/adapter-creation?step=2#versioning-and-releasing-your-adapter 
- [x] update https://docs.getdbt.com/docs/dbt-versions/core#how-we-version-adapter-plugins
- [x] update https://docs.getdbt.com/guides/adapter-creation?step=4#about-the-testing-framework
- [x] update https://docs.getdbt.com/guides/adapter-creation?step=4#testing-this-adapter 
- [x] update https://docs.getdbt.com/guides/adapter-creation?step=7#verify-a-new-adapter
- [x] Check w Amy: https://docs.getdbt.com/guides/adapter-creation?step=7#verified-by-dbt-labs
- [x] update https://docs.getdbt.com/guides/adapter-creation?step=8#release-cadence 
- [x] update diagram https://docs.getdbt.com/guides/adapter-creation?step=1#adapter-architecture 

### notes
For more context, see [slack thread](https://dbt-labs.slack.com/archives/C0131TY7EEA/p1710433062852709?thread_ts=1710193700.237319&cid=C0131TY7EEA).

### PR DESCRIPTION
**Pull request by nghi-ly**
## What are you changing in this pull request and why?

Update the adapter creation guide for the May 2 launch. 

closes #5104

## Checklist
- [x] Review the [Content style guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md) so my content adheres to these guidelines.
- [x] For [docs versioning](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#about-versioning), review how to [version a whole page](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version) and [version a block of content](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-blocks-of-content).
- [x] For any deleted sections, check to see if links need to be updated and if redirects need to be added where appropriate
- [x] For any deleted screenshots, remove image file if this is the last reference to it
- [x] Needs review from eng

### CODE REVIEW

File: website/docs/guides/adapter-creation.md [hunk 1]
```diff
@@ -1239,8 +1241,9 @@ Breaking this down:
...
     <Lightbox src="/img/adapter-guide/6-thank-contribs.png" title="thank yous"/>
 
+## Build a trusted adapter
 
-## Verify a new adapter
+The Trusted Adapter Program exists to allow adapter maintainers to demonstrate to the dbt community that your adapter is trusted to be used in production.
 
 The very first data platform dbt supported was Redshift followed quickly by Postgres (([dbt-core#174](https://github.com/dbt-labs/dbt-core/pull/174)). In 2017, back when dbt Labs (née Fishtown Analytics) was still a data consultancy, we added support for Snowflake and BigQuery. We also turned dbt's database support into an adapter framework ([dbt-core#259](https://github.com/dbt-labs/dbt-core/pull/259/)), and a plugin system a few years later. For years, dbt Labs specialized in those four data platforms and became experts in them. However, the surface area of all possible databases, their respective nuances, and keeping them up-to-date and bug-free is a Herculean and/or Sisyphean task that couldn't be done by a single person or even a single team! Enter the dbt community which enables dbt Core to work on more than 30 different databases (32 as of Sep '22)!
```
Comment (nghi-ly) [inline website/docs/guides/adapter-creation.md]: @dataders : if you could review the "trusted adapter" parts for accuracy, that would be great! 

changes per my chat with amy:
- moved the intro/historical content in "Verify a new adapter" to "Build a trusted adapter"
- deleted "Verify a new adapter"
- deleted other "verify | verified" subsections
- removed/replaced all references to "verified" in guide

File: website/docs/guides/adapter-creation.md [hunk 2]
```diff
@@ -106,9 +106,9 @@ A set of *materializations* and their corresponding helper macros defined in dbt
 ### Adapter Architecture
 
 
-Below is a diagram of how dbt-postgres, the adapter at the center of dbt-core, works.
+Below is a diagram of how the adapter architecture works.
```
Comment (colin-k-rogers) [inline website/docs/guides/adapter-creation.md]:
```suggestion
Below is a flow diagram illustrating how a `dbt run` command works with the `dbt-postgres` adapter. It shows the relationship between `dbt-core`, `dbt-adapters`, and individual adapters.
```

File: website/docs/guides/adapter-creation.md [hunk 3]
```diff
@@ -148,18 +148,20 @@ When your adapter becomes more popular, and people start using it, you may quick
 
 #### Keeping the code compatible with dbt Core
 
-New minor version releases of `dbt-core` may include changes to the Python interface for adapter plugins, as well as new or updated test cases. The maintainers of `dbt-core` will clearly communicate these changes in documentation and release notes, and they will aim for backwards compatibility whenever possible.
+An adapter is compatible with dbt Core if it has correctly implemented the interface defined in `dbt-adapters` and the tests defined in `dbt-tests-adapters`. Prior to dbt Core version 1.8, this interface was contained in `dbt-core`.
```
Comment (colin-k-rogers) [inline website/docs/guides/adapter-creation.md]:
```suggestion
An adapter is compatible with dbt Core if it has correctly implemented the interface defined in [dbt-adapters](https://github.com/dbt-labs/dbt-adapters/) and is tested by [dbt-tests-adapters](https://github.com/dbt-labs/dbt-adapters/tree/main/dbt-tests-adapter). Prior to dbt Core version 1.8, this interface was contained in `dbt-core`. 
```

File: website/docs/guides/adapter-creation.md [hunk 4]
```diff
@@ -148,18 +148,20 @@ When your adapter becomes more popular, and people start using it, you may quick
...
 
-Patch releases of `dbt-core` will _not_ include breaking changes to adapter-facing code. For more details, see ["About dbt Core versions"](/docs/dbt-versions/core).
+New minor version releases of `dbt-adapters` may include changes to the Python interface for adapter plugins, as well as new or updated test cases. The maintainers of `dbt-adapters` will clearly communicate these changes in documentation and release notes, and they will aim for backwards compatibility whenever possible.
+
+Patch releases of `dbt-adapters` will _not_ include breaking changes or new features to adapter-facing code.
 
 #### Versioning and releasing your adapter
 
-We strongly encourage you to adopt the following approach when versioning and releasing your plugin:
+dbt Labs strongly recommends you to adopt the following approach when versioning and releasing your plugin. Prior to dbt Core version 1.8, we recommended that the minor version of your plugin should match the minor version in `dbt-core` (e.g. 1.1.x).
```
Comment (colin-k-rogers) [inline website/docs/guides/adapter-creation.md]: > Prior to dbt Core version 1.8, 

Can we make this a callout and move to the bottom of the section?

File: website/docs/guides/adapter-creation.md [hunk 5]
```diff
@@ -148,18 +148,20 @@ When your adapter becomes more popular, and people start using it, you may quick
...
 #### Versioning and releasing your adapter
 
-We strongly encourage you to adopt the following approach when versioning and releasing your plugin:
+dbt Labs strongly recommends you to adopt the following approach when versioning and releasing your plugin. Prior to dbt Core version 1.8, we recommended that the minor version of your plugin should match the minor version in `dbt-core` (e.g. 1.1.x). 
 
-- The minor version of your plugin should match the minor version in `dbt-core` (e.g. 1.1.x).
-- Aim to release a new version of your plugin for each new minor version of `dbt-core` (once every three months).
-- While your plugin is new, and you're iterating on features, aim to offer backwards compatibility and deprecation notices for at least one minor version. As your plugin matures, aim to leave backwards compatibility and deprecation notices in place until the next major version (dbt Core v2).
-- Release patch versions of your plugins whenever needed. These patch releases should contain fixes _only_.
+- Declare major version compatibility with `dbt-adapters` and only set a boundary on the minor version if there is some known reason.
```
Comment (colin-k-rogers) [inline website/docs/guides/adapter-creation.md:161]: Can we add a bullet:
* Do not import/rely on code from `dbt-core`

**Review by colin-rogers-dbt — APPROVED** [review]

File: website/docs/docs/dbt-versions/core-versions.md [hunk 6]
```diff
@@ -99,24 +99,27 @@ We are committed to avoiding breaking changes in minor versions for end users of
 
 When you use dbt, you use a combination of `dbt-core` and an adapter plugin specific to your database. You can see the current list in [Supported Data Platforms](/docs/supported-data-platforms). Both `dbt-core` and dbt adapter plugins follow semantic versioning.
 
-`dbt-core` and adapter plugins coordinate new features and behind-the-scenes changes in minor releases. When fixing bugs, sooner is better, so patch versions are released independently for `dbt-core` and plugins.
+`dbt-core` and adapter plugins coordinate new features and behind-the-scenes changes via the `dbt-adapters` interface. New features in adapters, and that `dbt-core` will leverage, are defined in `dbt-adapters`. These features are opt-in so they only impact adapters that explicitly implement them. This allows us to independently release adapters, `dbt-adapters`, and `dbt-core` without creating a broken experience for users.
```
Comment (mirnawong1) [inline website/docs/docs/dbt-versions/core-versions.md]: this is great! i struggled to understand first time reading so tried to simplify
```suggestion
`dbt-core` and adapter plugins use the `dbt-adapters` interface to coordinate new features and behind-the-scenes changes. New adapter features are defined in `dbt-adapters` (which `dbt-core` will use). These features are opt-in, meaning they only impact adapters that explicitly implement them. This allows us to independently release adapters, `dbt-adapters`, and `dbt-core` without creating a broken experience for users.
```

File: website/docs/docs/dbt-versions/core-versions.md [hunk 7]
```diff
@@ -99,24 +99,27 @@ We are committed to avoiding breaking changes in minor versions for end users of
...
 
-That means that patch version numbers will likely differ between `dbt-core` and the adapter plugin(s) you have installed. However, major and minor version numbers should always match.
+That means that (unlike core versions prior to 1.8) the minor and patch version numbers will likely differ between `dbt-core` and the adapter plugin(s) you have installed.
```
Comment (mirnawong1) [inline website/docs/docs/dbt-versions/core-versions.md]:
```suggestion
Unlike `dbt-core` versions before 1.8, the minor and patch version numbers might not match between `dbt-core` and the adapter plugin(s) you've installed. 
```

File: website/docs/docs/dbt-versions/core-versions.md [hunk 8]
```diff
@@ -99,24 +99,27 @@ We are committed to avoiding breaking changes in minor versions for end users of
...
+`dbt-core` and adapter plugins coordinate new features and behind-the-scenes changes via the `dbt-adapters` interface. New features in adapters, and that `dbt-core` will leverage, are defined in `dbt-adapters`. These features are opt-in so they only impact adapters that explicitly implement them. This allows us to independently release adapters, `dbt-adapters`, and `dbt-core` without creating a broken experience for users.
 
-That means that patch version numbers will likely differ between `dbt-core` and the adapter plugin(s) you have installed. However, major and minor version numbers should always match.
+That means that (unlike core versions prior to 1.8) the minor and patch version numbers will likely differ between `dbt-core` and the adapter plugin(s) you have installed. 
+
+For example, you may find you're using `dbt-core==1.8.0` with `dbt-snowflake==1.9.0`. Even though these do not have the same minor version, they can still work together as they both work with `dbt-adapters==1.8.0`. Patch releases can contain important bug or security fixes so it’s critical to stay up to date.
```
Comment (mirnawong1) [inline website/docs/docs/dbt-versions/core-versions.md]: keeping it consistent
```suggestion
For example, you might find you're using `dbt-core==1.8.0` with `dbt-snowflake==1.9.0`. Even though these don't have the same minor version, they can still work together as they both work with `dbt-adapters==1.8.0`. Patch releases can contain important bug or security fixes so it’s critical to stay up to date. 
```

File: website/docs/guides/adapter-creation.md [hunk 9]
```diff
@@ -106,9 +106,9 @@ A set of *materializations* and their corresponding helper macros defined in dbt
 ### Adapter Architecture
 
 
-Below is a diagram of how dbt-postgres, the adapter at the center of dbt-core, works.
+Below is a flow diagram illustrating how a `dbt run` command works with the `dbt-postgres` adapter. It shows the relationship between `dbt-core`, `dbt-adapters`, and individual adapters.
```
Comment (mirnawong1) [inline website/docs/guides/adapter-creation.md:109]: for accessibility
```suggestion
The following flow diagram illustrates how a `dbt run` command works with the `dbt-postgres` adapter. It shows the relationship between `dbt-core`, `dbt-adapters`, and individual adapters.
```

File: website/docs/guides/adapter-creation.md [hunk 10]
```diff
@@ -148,18 +148,27 @@ When your adapter becomes more popular, and people start using it, you may quick
...
+
+- Declare major version compatibility with `dbt-adapters` and only set a boundary on the minor version if there is some known reason.
+- Do not import or rely on code from `dbt-core`. 
+- Aim to release a new minor version of your plugin as you add substantial new features. Typically, this will be triggered by adding support for new features released in `dbt-adapters` or by changes to the data platform itself.
+- While your plugin is new and you're iterating on features, aim to offer backwards compatibility and deprecation notices for at least one minor version. As your plugin matures, aim to leave backwards compatibility and deprecation notices in place until the next major version (dbt Core v2).
+- Release patch versions of your plugins whenever needed. These patch releases should only contain fixes.
+
+:::note
+
+Prior to dbt Core version 1.8, we recommended that the minor version of your plugin should match the minor version in `dbt-core` (e.g. 1.1.x).
```
Comment (mirnawong1) [inline website/docs/guides/adapter-creation.md]:
```suggestion
Prior to dbt Core version 1.8, we recommended that the minor version of your plugin should match the minor version in `dbt-core` (for example, 1.1.x).
```

**Review by mirnawong1 — APPROVED** [review]
looks great and approving again on our side! small suggestions on my end but they're not blockers!

File: website/docs/docs/dbt-versions/core-versions.md (hunk 6 above)
Comment (nghi-ly) [reply website/docs/docs/dbt-versions/core-versions.md]: thank you! like the suggs. adjusted it a bit to remove anthropomorphism

**Pull request merged**